internal/seedengine/seedengine.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2024 Edgeless Systems GmbH
+// SPDX-License-Identifier: AGPL-3.0-only
+
+// Package seedengine provides deterministic key derivation of ECDSA and symmetric keys
+// from a secret seed.
+package seedengine
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"io"
+
+	"filippo.io/keygen"
+	"golang.org/x/crypto/hkdf"
+)
+
+// SeedEngine provides deterministic key derivation of ECDSA and symmetric keys
+// from a secret seed.
+type SeedEngine struct {
+	curve   func() elliptic.Curve
+	hashFun func() hash.Hash
+	salt    []byte
+
+	podStateSeed []byte
+	historySeed  []byte
+
+	rootCAKey             *ecdsa.PrivateKey
+	transactionSigningKey *ecdsa.PrivateKey
+}
+
+// New creates a new SeedEngine from a secret seed and a salt.
+func New(secretSeed []byte, salt []byte) (*SeedEngine, error) {
+	se := &SeedEngine{
+		curve:   elliptic.P384,
+		hashFun: sha256.New,
+		salt:    salt,
+	}
+
+	// Recommended to use salt length equal to hash size, see RFC 5869, section 3.1.
+	if len(salt) != se.hashFun().Size() {
+		return nil, fmt.Errorf("salt must be %d bytes long", se.hashFun().Size())
+	}
+
+	var err error
+	se.podStateSeed, err = se.hkdfDerive(secretSeed, "POD STATE SECRET")
+	if err != nil {
+		return nil, fmt.Errorf("deriving seed: %w", err)
+	}
+	se.historySeed, err = se.hkdfDerive(secretSeed, "HISTORY SECRET")
+	if err != nil {
+		return nil, fmt.Errorf("deriving seed: %w", err)
+	}
+	transactionSigningSeed, err := se.hkdfDerive(secretSeed, "TRANSACTION SIGNING SECRET")
+	if err != nil {
+		return nil, fmt.Errorf("deriving seed: %w", err)
+	}
+	rootCASeed, err := se.hkdfDerive(secretSeed, "ROOT CA SEED")
+	if err != nil {
+		return nil, fmt.Errorf("deriving seed: %w", err)
+	}
+
+	se.transactionSigningKey, err = se.generateECDSAPrivateKey(transactionSigningSeed)
+	if err != nil {
+		return nil, fmt.Errorf("generating ECDSA key: %w", err)
+	}
+	se.rootCAKey, err = se.generateECDSAPrivateKey(rootCASeed)
+	if err != nil {
+		return nil, fmt.Errorf("generating ECDSA key: %w", err)
+	}
+
+	return se, nil
+}
+
+// DerivePodSecret derives a secret for a pod from the policy hash and the secret seed.
+func (s *SeedEngine) DerivePodSecret(policyHash []byte) ([]byte, error) {
```

Comment:
Not sure how relevant this is in practice, but should we check the length of the input here?

Reply:
I've switched the input type from `[]byte` to `[32]byte` and check explicitly for an empty array as well as for the hash of an empty byte slice.